### ISSUE TITLE
[Apollo] Weird camera noise smoothing and pattern

### ISSUE THREAD

**Issue by AntonioBiscuit** [sonyxperiadev/bug_tracker#620]
Platform: Tama
Device: Apollo
Kernel version: 4.14
Android version: PixelExperience_h8324-10.0-20200714-1039-OFFICIAL
Software binaries version: V9C

**Description**
There is strong and irregular noise smoothing at some point. 
Keeping the phone still or having similarities frame to frame will result in smoothing of those areas, while moving the phone fast enough, thus, having no similarities frame to frame, will show the noise as it really is in those areas.
It is the most noticeable on the edges of the preview/video in most cases and will appear in random areas.
The higher the ISO, the stronger the effect is overall.
Happens regardless of the other settings.

**How to reproduce**
Use phone with high enough ISO. Keep the phone still to see smoothing and weird stuff happen. You can also move it to show the "true noise" and more weird stuff.
You don't necessarilly need to record video, it is noticeable on the preview.

**Before showing you what it looks like**
Please note that the noise does appear different in dark areas and light areas. It feels "lower resolution"  in dark areas and full resolution in light areas. You can see some slight horizontal noise lines too. Both of these are things that I have always seen in stock Android and are not the issue.

Example of a frame showing the aforementionned "noise lines"

![image](https://user-images.githubusercontent.com/64173083/87667022-f4c0f000-c769-11ea-9fda-737d014e31ab.png)

**What it looks like**
Recorded using Hedge Cam, Bitrate 100Mbit/s, 2160p, ISO 8500, shutter 1/150s:

First video frame:
Noise is regular and is the same everywhere. This is the usual sensor noise at 8500 ISO when denoiser and all are disabled. Any frame of a video shot on the device should look like this without post-processing (as in stock)

![image](https://user-images.githubusercontent.com/64173083/87664693-fa1c3b80-c765-11ea-9ef8-2b002caee309.png)

A few frames later:
Noise is irregular. It is smoothed down in a big part of the image but not in some areas, which shows the "true noise" and makes weird fuss appear on the image. This looks very bad in my opinion.
You can also see that the bottom and left edges (in this frame) have a lot of noise. You can see this on any edge depending on the moment. This is recurrent and is the area where the "true noise" will tend to stay compared to other areas.

![image](https://user-images.githubusercontent.com/64173083/87665271-f1783500-c766-11ea-98e8-501807998a75.png)

Settings
Every denoiser/post-processing related setting is disabled. (There are more settings in Hedge Cam than in Open Camera)

![Screenshot_20200716-125019](https://user-images.githubusercontent.com/64173083/87668498-73b72800-c76c-11ea-8578-0cdaf8a0d74b.png)

**Comment by sjllls** [comment]
I noticed this since I was working on Yoshino platform....But it should not be a bug, because IMX400 has a lot of noise itself. We had the same symptom on stock based LineageOS while using Snap. Sony provides strong and perfect noise reducing algorithm on SEMC camera. So you could get nice picture with stock camera. But SODP provides full camera 2 API for you including RAW support. So try a better camera App or avoiding using camera in such a dark environment.

**Comment by AntonioBiscuit** [comment]
The noise from IMX 400 itself is not a big deal for me, I'm used to it. 

The problem really is the smoothing pattern itself, it's behaviour. 
At this very moment, I would rather have no noise reduction and smoothing at all since it is blocky. 

I consider this as a bug since there are blocks of the frame which are denoised and others which are not. It just nasty to me. 

Even if you disable all denoiser options in your camera settings, this blocky denoiser which I talk about is always present and can't be disabled.

However, IF it worked flawlessly without blocking artifacts on the full frame, I would be glad and would even leave it ON to help a little bit with video compression (if you could disable it anyway).

Not being able to use your camera in dim environnements while you could do it before is a bit of a bummer to me.
It can behave very good if you know how to use it, even in dark areas in my opinion.

I can't think of any video/camera app that has given me better results than Open Camera or Hedge Cam in the past.
If you know any better alternative, feel free to share :)

If you know where or when in the video process this noise reduction happens (or how to disable it with a command or something) I would be happy to know :)

**Comment by sjllls** [comment]
> The noise from IMX 400 itself is not a big deal for me, I'm used to it.
....
> If you know where or when in the video process this noise reduction happens (or how to disable it with a command or something) I would be happy to know :)

Sorry for misunderstand, you are talking about these blocks. You are right, there should be something wrong with CAM drivers, There blocks are obvious when using Gcam Night Sight ( I always think it's a bug of Gcam on IMX400 ). But this issue exists for a long time since Yoshino devices has been added to SODP..

**Comment by mhtvsSFrpHdE** [comment]
I rather prefer to think there is a bug. But SODP has no responsibility to provide a correct camera,  
or I would say a camera bad quality like this is "by designed".

Let's say you've got Nvidia 2080 Ti video cards, and now you throw those Windows system and driver away,  
instead, you decide to use Linux and open source driver...

You don't deserve to get anything else than a basic video output,  no matter what hardware really is.

**Comment by sraase** [comment]
The IMX400 is a quite complicated sensor. For in-sensor noise reduction, the sensor supports special low-noise ADC modes for low-light situations, which cause saturation in normal light. On Open Devices, these modes are not used, so I expect worse low-light performance compared to Xperia software. Instead, Open Devices uses in-sensor noise reduction to compensate (which is not used on Xperia).

Basically, Open Devices uses a static sensor configuration with a somewhat default IQ tuning on Qualcomm's software, while Xperia uses a more dynamic sensor configuration and different algorithms.

**Comment by mikkorantalainen** [comment]
> Basically, Open Devices uses a static sensor configuration with a somewhat default IQ tuning on Qualcomm's software, while Xperia uses a more dynamic sensor configuration and different algorithms.

Is it possible to tweak the static tuning somehow?

**Comment by jerpelea** [comment]
unfortunately no.

**Comment by mhtvsSFrpHdE** [comment]
That sounds like if you can edit some system configuration and reboot to perform a switch to turn on one of the modes.

**Comment by sraase** [comment]
You can partially influence the behaviour through the standard Android Camera2 API.
But the sensor configuration and tuning data are fix. There is no system configuration file to edit except for the driver itself.

**Comment by mhtvsSFrpHdE** [comment]
@sraase Yes, but everything would finally compile to binary. They are maybe a few of `.so` files out there, we can edit source code, make several compile result then pack them to different zip, to flash them and get changed sensor configuration and tuning data.

At compile time this is a very uneconomical behavior in time, however people like me may not care left the PC turn on a night

The rest problem would be what to edit, and edit to what.

Well although that's sounds like we are talking about how to tune a classic car's manufacturer hidden engine secrets. We probably asked too much. Would be very impolite.

**Comment by sraase** [comment]
You _could_ change the driver and tuning data, if you had access to the source and tools.
Compilation time is not an issue - reproducable environments, time, effort and subjectivity are.
Unfortunately.

**Comment by sjllls** [comment]
@kgvarunkanth gave us a solution.
Please try Gcam : [Link](https://www.celsoazevedo.com/files/android/google-camera/f/changelog1518/) with this xml config.
[new trcamera imx400 dark.zip](https://github.com/sonyxperiadev/bug_tracker/files/6240843/new.trcamera.imx400.dark.zip)

Some example picture:
![photo_2021-04-01_06-55-01](https://user-images.githubusercontent.com/29087038/113240746-df7ff780-92df-11eb-9e57-176539e29333.jpg)
![photo_2021-04-01_11-46-47](https://user-images.githubusercontent.com/29087038/113240787-f4f52180-92df-11eb-9b3c-0124bfcfd619.jpg)

**Comment by kgvarunkanth** [comment]
Did anyone test the config file?
